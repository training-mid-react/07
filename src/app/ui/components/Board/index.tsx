import { useState } from 'react';
import { BoardMatrix, Player } from '../../../core/interfaces/board-state';
import './style.scss';

interface Props{
  board: BoardMatrix,
  handleSelectColumn: (column: number) => void
  currentPlayer: Exclude<Player, 'draw'>
}
export const Board = ({board, handleSelectColumn, currentPlayer}: Props) => {
  
  const [animatedCell, setAnimatedCell] = useState<{ col: number; row: number } | null>(null);

  const handleCellClick = (colIndex: number) => {
    const rowIndex = board.findIndex(row => row[colIndex] === null);
    if (rowIndex !== -1) {
      setAnimatedCell({ col: colIndex, row: rowIndex });
      setTimeout(() => {
        handleSelectColumn(colIndex);
        setAnimatedCell(null);
      }, 500); 
    }
  };

  return (
    
    <div className="board">
      {
        board.map((row, rowIndex) => {
          return (
            <div 
              key={`row_${rowIndex}`} 
              className='board__row'
            >
              {row.map((cell, colIndex) => {
                const isFalling = animatedCell?.col === colIndex && animatedCell.row >= rowIndex;
                const fallDistance = isFalling ? `${50 * (animatedCell!.row - rowIndex)}px` : '0px';
                const animationFall = {
                  transform: `translateY(${fallDistance})`,
                };
                return( 
                  <div 
                    key={`cell_${colIndex}`} 
                    data-testid={`cell_${rowIndex}_${colIndex}`}
                    className={`board__col ${cell || ''} ${isFalling ? `falling--${currentPlayer}` : ''}`}  
                    onClick={() => handleCellClick(colIndex)}
                    style={isFalling ? animationFall : {}}
                  />
                );
              })}
            </div>
          );
        })
      }
      
    </div>
  );
};
