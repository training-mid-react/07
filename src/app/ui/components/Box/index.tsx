import { CELL } from "@/app/core/interfaces/game.interface";
import "./style.scss"

interface Props {
    value: CELL;
    onClick: () => void;
}

const Box = ({ value, onClick }: Props) => {
    return (
        <div data-testid="box" className="box" onClick={onClick}>
            {
                value && <div style={{backgroundColor: value.color}} className="box__chip"></div>
            }
        </div>
    );
}

export default Box;