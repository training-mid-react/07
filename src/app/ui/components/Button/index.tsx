import React from 'react';
import './style.scss';

interface ButtonProps {
  children: React.ReactNode;
  className?: string;
  onClick: () => void;
  disabled?: boolean;
}

const Button: React.FC<ButtonProps> = ({ children, onClick, disabled, className }) => {
  return (
    <button data-testid='button-component' className={className || 'custom-button'} onClick={onClick} disabled={disabled}>
      {children}
    </button>
  );
};

export default Button;
