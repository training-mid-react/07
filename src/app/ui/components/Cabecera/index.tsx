
import "./style.scss";

interface Props {
    currentPlayer: number
}

const Cabecera = ({ currentPlayer }: Props) => {
    return (
        <>
            <section className="cabecera">
                <div>
                    <h2>Jugador 1</h2>
                    <div className="ficha-cabecera ficha-cabecera--jugador1"></div>
                </div>
                <div>
                    <h2>Jugador 2</h2>
                    <div className="ficha-cabecera ficha-cabecera--jugador2"></div>
                </div>
            </section>
            <h3>Turno del jugador: {currentPlayer}</h3>
        </>

    )
}

export default Cabecera 