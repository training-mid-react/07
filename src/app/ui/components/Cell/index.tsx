import { FC } from 'react';
import { PLAYER } from '../../../../core/interfaces';
import './style.scss';

export interface CellProps {
  value: PLAYER | null;
}

export const Cell: FC<CellProps> = ({ value }) => {

  return (
    <>
    <div className={`cell cell__${value}`} />
    <div className='cell__frame'></div>
    </>
  );
};
