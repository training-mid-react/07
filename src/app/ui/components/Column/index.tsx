import './style.scss';
import { useDropCell } from '../../../core/hooks/useDropCell';

interface ColumnProps {
  columnIndex: number;
  dropPiece: (column: number) => void;
  columnData: number[];
}

function Column({ columnIndex, dropPiece, columnData }: ColumnProps) {
  const { handleColumnClick, droppingCell } = useDropCell({ columnIndex, dropPiece, columnData })
  return (
    <div className="column" onClick={handleColumnClick}>
      {columnData.map((cell, rowIndex) => {
        const isDropping = droppingCell === rowIndex;
        const cellClass = `cell ${cell === 1 ? 'cell--player1' : cell === 2 ? 'cell--player2' : ''} ${isDropping ? 'cell--dropping' : ''}`;
        return (
          <div
            key={rowIndex}
            className={cellClass}
          />
        );
      })}
    </div>
  );
};

export default Column;
