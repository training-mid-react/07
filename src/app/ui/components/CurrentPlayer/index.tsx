import './style.scss';

interface ICurrentPlayerProps {
  currentPlayer: number;
}

export const CurrentPlayer = ({ currentPlayer }: ICurrentPlayerProps) => {
  return (
    <div className="current-player">
      Es el turno del Jugador: 
      <span className={currentPlayer === 1 ? 'current-player__player1' : 'current-player__player2'}>
        {currentPlayer}
      </span>
    </div>
  );
};
