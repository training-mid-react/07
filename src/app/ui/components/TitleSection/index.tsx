
import './style.scss';

interface Props {
  contentTitle: string
}
export const TitleSection = ({contentTitle}: Props) => {

  return (
    <h2 className="title-section">{contentTitle}</h2>
  );
};
