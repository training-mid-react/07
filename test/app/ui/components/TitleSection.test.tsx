import React from 'react';

import { render } from '@testing-library/react';
import {TitleSection} from '../../../../src/app/ui/components/TitleSection';

describe('TitleSection', () => {

  // ===> Initial state
  const contentTitle = 'title';

  test('Matches the snapshot', () => {
    const {asFragment} = render(
      <TitleSection 
        contentTitle={contentTitle} 
      />
    );
  
    expect(asFragment()).toMatchSnapshot();
  });

  test('Expect to render the title', () => {
    const {getByText} = render(
      <TitleSection 
        contentTitle={contentTitle} 
      />
    );
    
    expect(getByText(contentTitle)).toBeDefined();
  });
  
});