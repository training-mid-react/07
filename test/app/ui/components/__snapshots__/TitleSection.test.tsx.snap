// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TitleSection > Matches the snapshot 1`] = `
<DocumentFragment>
  <h2
    class="title-section"
  >
    title
  </h2>
</DocumentFragment>
`;
